In [13]:
import pandas as pd

In [14]:
# Read SUBMISSION file into a DataFrame
submission = pd.read_csv(
    'raw_data/2026q1/SUBMISSION.tsv',
    sep='\t', # separator '\t' tells pandas the columns are separated by tabs, not commas (confirmed in Phase 1)
    dtype={'ISSUERCIK': str} # dtype forces specified column to stay as text (str), protecting leading zeros that a number would silently drop
)

# Quick sanity check: how many rows, and what do the first 5 look like?
print(submission.shape)
submission.head()

(69259, 14)


,ACCESSION_NUMBER,FILING_DATE,PERIOD_OF_REPORT,DATE_OF_ORIG_SUB,NO_SECURITIES_OWNED,NOT_SUBJECT_SEC16,FORM3_HOLDINGS_REPORTED,FORM4_TRANS_REPORTED,DOCUMENT_TYPE,ISSUERCIK,ISSUERNAME,ISSUERTRADINGSYMBOL,REMARKS,AFF10B5ONE
0,0001193125-26-134840,31-MAR-2026,27-MAR-2026,NaN,NaN,false,NaN,NaN,4,0001825079,"Velo3D, Inc.",VELO,NaN,false
1,0002060688-26-000006,31-MAR-2026,30-MAR-2026,NaN,NaN,0,NaN,NaN,4,0001874315,Satellogic Inc.,SATL,NaN,0
2,0001104659-26-037685,31-MAR-2026,18-MAR-2026,NaN,0.0,NaN,NaN,NaN,3,0001969302,Pony AI Inc.,PONY,Exhibit 24 - Power of Attorney,NaN
3,0000895345-26-000136,31-MAR-2026,18-MAR-2026,NaN,1.0,NaN,NaN,NaN,3,0000912892,"GRUPO TELEVISA, S.A.B.",TV,NaN,NaN
4,0002123440-26-000001,31-MAR-2026,18-MAR-2026,NaN,1.0,NaN,NaN,NaN,3,0002074123,CCH Holdings Ltd,CCHH,NaN,NaN


In [15]:
# Read REPORTINGOWNER.tsv into a DataFrame.
# sep='\t' tells pandas the columns are tab-separated, not comma-separated.
# dtype forces RPTOWNERCIK to stay text, protecting leading zeros in this CIK field.
reportingowner = pd.read_csv(
    'raw_data/2026q1/REPORTINGOWNER.tsv',
    sep='\t',
    dtype={'RPTOWNERCIK': str}
)

print(reportingowner.shape)
reportingowner.head()

(72211, 13)


,ACCESSION_NUMBER,RPTOWNERCIK,RPTOWNERNAME,RPTOWNER_RELATIONSHIP,RPTOWNER_TITLE,RPTOWNER_TXT,RPTOWNER_STREET1,RPTOWNER_STREET2,RPTOWNER_CITY,RPTOWNER_STATE,RPTOWNER_ZIPCODE,RPTOWNER_STATE_DESC,FILE_NUMBER
0,0001493152-26-014456,0002125911,Ong Sie Hou Raymond,Director,NaN,NaN,"600 NORTH BRIDGE ROAD,",PARKVIEW SQUARE #24-01,SINGAPORE,NaN,188778,SINGAPORE,001-42426
1,0001213900-26-037919,0002125166,Huang Huaguang,Officer,Chief Financial Officer,NaN,"ROOM 1, BUILDING 10, JINZHONG GUOBIN HUI","QINYUAN 2ND ROAD, XILI STREET",SHENZHEN,NaN,518071,CHINA,001-33768
2,0001193125-26-135334,0001238734,WALKER JOHN KENT,Officer,"President, Global Affairs, CLO",NaN,C/O ALPHABET INC.,1600 AMPHITHEATRE PKWY,MOUNTAIN VIEW,CA,94043,NaN,001-37580
3,0001030471-26-000005,0001432657,Shao Sean,Director,NaN,NaN,"4TH FLOOR, SOUTH WING","368 LIUHE ROAD, BINJIANG DISTRICT",HANGZHOU,NaN,310053,CHINA,001-35216
4,0001062993-26-001774,0002122229,Tavares Craig,Officer,BUZZ HPC- President & COO,NaN,"7900 CALLAGHAN ROAD, SUITE 128",NaN,SAN ANTONIO,TX,78229,NaN,001-40398


In [16]:
# Read NONDERIV_TRANS.tsv into a DataFrame — same pattern as your last 2 reads
nonderiv_trans = pd.read_csv(
    'raw_data/2026q1/NONDERIV_TRANS.tsv',
    sep='\t'
)

print(nonderiv_trans.shape)
nonderiv_trans.head()

(103733, 28)


C:\Users\stephanie.mitchell\AppData\Local\Temp\ipykernel_15272\2065309264.py:2: DtypeWarning: Columns (0: VALU_OWND_FOLWNG_TRANS_FN) have mixed types. Specify dtype option on import or set low_memory=False.
  nonderiv_trans = pd.read_csv(


,ACCESSION_NUMBER,NONDERIV_TRANS_SK,SECURITY_TITLE,SECURITY_TITLE_FN,TRANS_DATE,TRANS_DATE_FN,DEEMED_EXECUTION_DATE,DEEMED_EXECUTION_DATE_FN,TRANS_FORM_TYPE,TRANS_CODE,...,TRANS_ACQUIRED_DISP_CD,TRANS_ACQUIRED_DISP_CD_FN,SHRS_OWND_FOLWNG_TRANS,SHRS_OWND_FOLWNG_TRANS_FN,VALU_OWND_FOLWNG_TRANS,VALU_OWND_FOLWNG_TRANS_FN,DIRECT_INDIRECT_OWNERSHIP,DIRECT_INDIRECT_OWNERSHIP_FN,NATURE_OF_OWNERSHIP,NATURE_OF_OWNERSHIP_FN
0,0001193125-26-135334,8937906,Class C Capital Stock,NaN,31-MAR-2026,NaN,NaN,NaN,4.0,G,...,A,NaN,60801.0,NaN,NaN,NaN,I,NaN,NaN,NaN
1,0001193125-26-135334,8937901,Class C Capital Stock,NaN,27-MAR-2026,NaN,NaN,NaN,4.0,S,...,D,NaN,56608.0,NaN,NaN,NaN,I,NaN,"By Arete Trust, John Kent Walker and Diana Rut...",NaN
2,0001193125-26-135334,8937905,Class C Capital Stock,NaN,31-MAR-2026,NaN,NaN,NaN,4.0,G,...,D,NaN,23508.0,NaN,NaN,NaN,D,NaN,"By Arete Trust, John Kent Walker and Diana Rut...",NaN
3,0001193125-26-135334,8937904,Class C Capital Stock,NaN,27-MAR-2026,NaN,NaN,NaN,4.0,S,...,D,NaN,51808.0,NaN,NaN,NaN,I,NaN,"By Arete Trust, John Kent Walker and Diana Rut...",NaN
4,0001193125-26-135334,8937900,Class C Capital Stock,NaN,27-MAR-2026,NaN,NaN,NaN,4.0,S,...,D,NaN,58124.0,NaN,NaN,NaN,I,NaN,"By Arete Trust, John Kent Walker and Diana Rut...",NaN


In [17]:
# Filter nonderiv_trans down to only Purchase (P) and Sale (S) transactions.

# .isin() checks TRANS_CODE against the list (P&S) and returns True/False per row.
# Outter nonderiv_trans [] is to keep any rows that return True (P & S) and anything else is dropped
nonderiv_trans = nonderiv_trans[nonderiv_trans['TRANS_CODE'].isin(['P', 'S'])]

print(nonderiv_trans.shape)
nonderiv_trans.head()

(28757, 28)


,ACCESSION_NUMBER,NONDERIV_TRANS_SK,SECURITY_TITLE,SECURITY_TITLE_FN,TRANS_DATE,TRANS_DATE_FN,DEEMED_EXECUTION_DATE,DEEMED_EXECUTION_DATE_FN,TRANS_FORM_TYPE,TRANS_CODE,...,TRANS_ACQUIRED_DISP_CD,TRANS_ACQUIRED_DISP_CD_FN,SHRS_OWND_FOLWNG_TRANS,SHRS_OWND_FOLWNG_TRANS_FN,VALU_OWND_FOLWNG_TRANS,VALU_OWND_FOLWNG_TRANS_FN,DIRECT_INDIRECT_OWNERSHIP,DIRECT_INDIRECT_OWNERSHIP_FN,NATURE_OF_OWNERSHIP,NATURE_OF_OWNERSHIP_FN
1,0001193125-26-135334,8937901,Class C Capital Stock,NaN,27-MAR-2026,NaN,NaN,NaN,4.0,S,...,D,NaN,56608.0,NaN,NaN,NaN,I,NaN,"By Arete Trust, John Kent Walker and Diana Rut...",NaN
3,0001193125-26-135334,8937904,Class C Capital Stock,NaN,27-MAR-2026,NaN,NaN,NaN,4.0,S,...,D,NaN,51808.0,NaN,NaN,NaN,I,NaN,"By Arete Trust, John Kent Walker and Diana Rut...",NaN
4,0001193125-26-135334,8937900,Class C Capital Stock,NaN,27-MAR-2026,NaN,NaN,NaN,4.0,S,...,D,NaN,58124.0,NaN,NaN,NaN,I,NaN,"By Arete Trust, John Kent Walker and Diana Rut...",NaN
5,0001193125-26-135334,8937903,Class C Capital Stock,NaN,27-MAR-2026,NaN,NaN,NaN,4.0,S,...,D,NaN,53108.0,NaN,NaN,NaN,I,NaN,"By Arete Trust, John Kent Walker and Diana Rut...",NaN
6,0001193125-26-135334,8937902,Class C Capital Stock,NaN,27-MAR-2026,NaN,NaN,NaN,4.0,S,...,D,NaN,55208.0,NaN,NaN,NaN,I,NaN,"By Arete Trust, John Kent Walker and Diana Rut...",NaN


In [18]:
# Standarize NOT_SUBJECT_SEC16 in the submission table to real Boolean values
submission['NOT_SUBJECT_SEC16'] = submission['NOT_SUBJECT_SEC16'].replace({
    'false' : False,
    0 : False,
    '0' : False,
    'true' : True,
    1 : True,
    '1' : True

})

submission['NOT_SUBJECT_SEC16'].unique()

array([False, nan, True], dtype=object)

In [19]:
submission['AFF10B5ONE'].unique()

<StringArray>
['false', '0', nan, '1', 'true']
Length: 5, dtype: str

In [20]:
# standardize AFF10B5ONE to real Boolean values
submission['AFF10B5ONE'] = submission['AFF10B5ONE'].replace({
    'false' : False,
    '0' : False,
    'true' : True,
    '1' : True
})

# Confirming changes
submission['AFF10B5ONE'].unique()

array([False, nan, True], dtype=object)

In [21]:
nonderiv_trans['EQUITY_SWAP_INVOLVED'].unique()

<StringArray>
['false', '0', '1']
Length: 3, dtype: str

In [22]:
nonderiv_trans['EQUITY_SWAP_INVOLVED'] = nonderiv_trans['EQUITY_SWAP_INVOLVED'].replace({
    'false' : False,
    '0': False,
    '1' : True
})

nonderiv_trans['EQUITY_SWAP_INVOLVED'].unique()

array([False, True], dtype=object)

## Cleaning: RPTOWNER_RELATIONSHIP (multi-role field)

SEC stores multiple insider roles as one comma-separated string 
(e.g. "Director,Officer").<br> 
Splitting into separate boolean flag 
columns so each role is independently queryable, without losing 
any role a person actually holds.

In [23]:
reportingowner['RPTOWNER_RELATIONSHIP'].unique()

<StringArray>
[                              'Director',
                                'Officer',
       'Director,Officer,TenPercentOwner',
                       'Director,Officer',
               'Director,TenPercentOwner',
                        'TenPercentOwner',
                                  'Other',
         'Director,TenPercentOwner,Other',
                 'Director,Officer,Other',
                          'Officer,Other',
                         'Director,Other',
                'Officer,TenPercentOwner',
                  'TenPercentOwner,Other',
 'Director,Officer,TenPercentOwner,Other',
          'Officer,TenPercentOwner,Other',
                                      nan]
Length: 16, dtype: str

In [24]:
# Verifying NAN for cleaning design
reportingowner['RPTOWNER_RELATIONSHIP'].isna().sum()

np.int64(19)

In [25]:
# Flag each role from RPTOWNER_RELATIONSHIP. Two steps per flag:
# 1) .str.contains() checks if the role name appears anywhere in the (possibly combined) text
# 2) .astype('boolean') converts to a nullable boolean type so missing relationships can stay NA instead of being forced to False
# Then .loc explicitly sets the True NA rows (no relationship listed at all) to NA, since step 1 alone doesn't reliably preserve that

reportingowner['is_officer'] = reportingowner['RPTOWNER_RELATIONSHIP'].str.contains('Officer').astype('boolean')
reportingowner.loc[reportingowner['RPTOWNER_RELATIONSHIP'].isna(), 'is_officer'] = pd.NA

reportingowner['is_director'] = reportingowner['RPTOWNER_RELATIONSHIP'].str.contains('Director').astype('boolean')
reportingowner.loc[reportingowner['RPTOWNER_RELATIONSHIP'].isna(), 'is_director'] = pd.NA

reportingowner['is_ten_percent_owner'] = reportingowner['RPTOWNER_RELATIONSHIP'].str.contains('TenPercentOwner').astype('boolean')
reportingowner.loc[reportingowner['RPTOWNER_RELATIONSHIP'].isna(), 'is_ten_percent_owner'] = pd.NA

reportingowner['is_other'] = reportingowner['RPTOWNER_RELATIONSHIP'].str.contains('Other').astype('boolean')
reportingowner.loc[reportingowner['RPTOWNER_RELATIONSHIP'].isna(), 'is_other'] = pd.NA

In [26]:
reportingowner[reportingowner['RPTOWNER_RELATIONSHIP'].isna()][['is_officer', 'is_director', 'is_ten_percent_owner', 'is_other']]

,is_officer,is_director,is_ten_percent_owner,is_other
6219,<NA>,<NA>,<NA>,<NA>
6940,<NA>,<NA>,<NA>,<NA>
6941,<NA>,<NA>,<NA>,<NA>
20821,<NA>,<NA>,<NA>,<NA>
22500,<NA>,<NA>,<NA>,<NA>
22505,<NA>,<NA>,<NA>,<NA>
23037,<NA>,<NA>,<NA>,<NA>
24541,<NA>,<NA>,<NA>,<NA>
24855,<NA>,<NA>,<NA>,<NA>
24856,<NA>,<NA>,<NA>,<NA>


## Cleaning: Dropping _FN footnote columns

_FN columns hold footnote reference codes (F1, F2, etc..) , not analytical <br>
None of the 5 business questions need footnote text, so dropping them from all 3 tables to simplify schema going forward.

In [29]:
# Confirm which columns end in _FN before dropping
fn_columns = nonderiv_trans.columns[nonderiv_trans.columns.str.endswith('_FN')]
print(fn_columns)

Index(['SECURITY_TITLE_FN', 'TRANS_DATE_FN', 'DEEMED_EXECUTION_DATE_FN',
       'EQUITY_SWAP_TRANS_CD_FN', 'TRANS_TIMELINESS_FN', 'TRANS_SHARES_FN',
       'TRANS_PRICEPERSHARE_FN', 'TRANS_ACQUIRED_DISP_CD_FN',
       'SHRS_OWND_FOLWNG_TRANS_FN', 'VALU_OWND_FOLWNG_TRANS_FN',
       'DIRECT_INDIRECT_OWNERSHIP_FN', 'NATURE_OF_OWNERSHIP_FN'],
      dtype='str')


In [ ]:
# Dropping columns from nonderiv_trans
nonderiv_trans = nonderiv_trans.drop(columns=fn_columns)

In [ ]:
fn_columns = submission.columns[submission.columns.str.endswith('_FN')]
print(fn_columns)

Index([], dtype='str')


In [38]:
submission = submission.drop(columns=fn_columns)

In [ ]:
fn_columns = reportingowner.columns[reportingowner.columns.str.endswith('_FN')]
print(fn_columns)

Index([], dtype='str')


In [40]:
reportingowner = reportingowner.drop(columns=fn_columns)

In [ ]:
# Check row count still the same
print(nonderiv_trans.shape)

(28757, 16)


In [ ]:
# Check for unexpected nulls 
print(nonderiv_trans[['TRANS_DATE', 'TRANS_SHARES', 'TRANS_PRICEPERSHARE']].isna().sum())

TRANS_DATE              0
TRANS_SHARES            0
TRANS_PRICEPERSHARE    58
dtype: int64


In [44]:
# Verify flags & booleans actually correct
print(reportingowner[['is_officer', 'is_director', 'is_ten_percent_owner', 'is_other']].dtypes)

is_officer              boolean
is_director             boolean
is_ten_percent_owner    boolean
is_other                boolean
dtype: object


In [ ]:
print(nonderiv_trans.columns.tolist())

['ACCESSION_NUMBER', 'NONDERIV_TRANS_SK', 'SECURITY_TITLE', 'TRANS_DATE', 'DEEMED_EXECUTION_DATE', 'TRANS_FORM_TYPE', 'TRANS_CODE', 'EQUITY_SWAP_INVOLVED', 'TRANS_TIMELINESS', 'TRANS_SHARES', 'TRANS_PRICEPERSHARE', 'TRANS_ACQUIRED_DISP_CD', 'SHRS_OWND_FOLWNG_TRANS', 'VALU_OWND_FOLWNG_TRANS', 'DIRECT_INDIRECT_OWNERSHIP', 'NATURE_OF_OWNERSHIP']


In [46]:
nonderiv_trans[nonderiv_trans['TRANS_PRICEPERSHARE'].isna()].head(10)

,ACCESSION_NUMBER,NONDERIV_TRANS_SK,SECURITY_TITLE,TRANS_DATE,DEEMED_EXECUTION_DATE,TRANS_FORM_TYPE,TRANS_CODE,EQUITY_SWAP_INVOLVED,TRANS_TIMELINESS,TRANS_SHARES,TRANS_PRICEPERSHARE,TRANS_ACQUIRED_DISP_CD,SHRS_OWND_FOLWNG_TRANS,VALU_OWND_FOLWNG_TRANS,DIRECT_INDIRECT_OWNERSHIP,NATURE_OF_OWNERSHIP
857,0001193125-26-132846,8938684,Common Stock,26-MAR-2026,NaN,4.0,P,False,NaN,37187.0,NaN,A,37547373.0,NaN,I,See footnotes
3887,0001493152-26-012483,8935028,Common Stock,20-MAR-2026,20-MAR-2026,4.0,P,False,NaN,9720.0,NaN,A,9720.0,NaN,I,Laurence Marton Rollover IRA
3888,0001493152-26-012482,8936095,Common Stock,20-MAR-2026,20-MAR-2026,4.0,P,False,NaN,24300.0,NaN,A,313357.0,NaN,I,The Bagai Family Trust
3889,0001493152-26-012481,8934722,Common Stock,20-MAR-2026,20-MAR-2026,4.0,P,False,NaN,97200.0,NaN,A,97200.0,NaN,I,Voll Family Trust DTD 6/23/2010 Mark Voll TTEE
4699,0001493152-26-012371,8934225,Class A ordinary shares,23-MAR-2026,NaN,4.0,P,False,NaN,275000.0,NaN,A,275000.0,NaN,I,See footnote
4700,0001493152-26-012369,8935812,Class A ordinary shares,23-MAR-2026,NaN,4.0,P,False,NaN,275000.0,NaN,A,275000.0,NaN,D,NaN
6143,0001213900-26-032495,8932422,Class A ordinary shares,20-MAR-2026,NaN,4.0,P,False,NaN,135000.0,NaN,A,135000.0,NaN,D,NaN
8300,0001213900-26-031142,8931408,Class A Ordinary Shares,16-MAR-2026,NaN,4.0,P,False,NaN,160000.0,NaN,A,160000.0,NaN,D,NaN
8301,0001213900-26-031141,8932404,Class A Ordinary Shares,16-MAR-2026,NaN,4.0,P,False,NaN,160000.0,NaN,A,160000.0,NaN,I,See Footnote
15291,0001099910-26-000120,8922994,Common Stock,13-MAR-2026,NaN,4.0,P,False,NaN,1250000.0,NaN,A,10178196.0,NaN,D,NaN


In [ ]:
# How many rows rows with a missing price 
nonderiv_trans[nonderiv_trans['TRANS_PRICEPERSHARE'].isna()]['TRANS_CODE'].value_counts()

TRANS_CODE
P    42
S    16
Name: count, dtype: int64

In [48]:
# Checking sale-coded missing price rows
nonderiv_trans[(nonderiv_trans['TRANS_PRICEPERSHARE'].isna()) & (nonderiv_trans['TRANS_CODE'] == 'S')].head(10)

,ACCESSION_NUMBER,NONDERIV_TRANS_SK,SECURITY_TITLE,TRANS_DATE,DEEMED_EXECUTION_DATE,TRANS_FORM_TYPE,TRANS_CODE,EQUITY_SWAP_INVOLVED,TRANS_TIMELINESS,TRANS_SHARES,TRANS_PRICEPERSHARE,TRANS_ACQUIRED_DISP_CD,SHRS_OWND_FOLWNG_TRANS,VALU_OWND_FOLWNG_TRANS,DIRECT_INDIRECT_OWNERSHIP,NATURE_OF_OWNERSHIP
16319,0001493152-26-009984,8921568,Common Stock,12-MAR-2026,NaN,4.0,S,False,NaN,25000.0,NaN,D,175000.0,NaN,D,NaN
16320,0001493152-26-009984,8921567,Common Stock,01-OCT-2025,NaN,4.0,S,False,NaN,17750.0,NaN,D,200000.0,NaN,D,NaN
16651,0001193125-26-104662,8921671,Class A Common Stock,09-MAR-2026,NaN,4.0,S,False,NaN,2350068.0,NaN,D,23717859.0,NaN,D,NaN
17285,0001493152-26-009896,8922157,Common Stock,11-MAR-2026,NaN,4.0,S,False,NaN,25000.0,NaN,D,687500.0,NaN,D,NaN
18883,0001193125-26-102576,8919774,Class A Common Stock,09-MAR-2026,NaN,4.0,S,False,NaN,2350068.0,NaN,D,23717859.0,NaN,D,NaN
38756,0001193125-26-086198,8899070,Series W-7B Variable Rate Demand Preferred Shares,23-FEB-2026,NaN,4.0,S,False,NaN,1464.0,NaN,D,0.0,NaN,I,By Subsidiary
43108,0001193125-26-078713,8896158,Series W-7 Variable Rate Muni Term Preferred S...,19-FEB-2026,NaN,4.0,S,False,NaN,777.0,NaN,D,0.0,NaN,I,By Subsidiary
43137,0001193125-26-078700,8895377,Series W-7 Variable Rate Muni Term Preferred S...,19-FEB-2026,NaN,4.0,S,False,NaN,1536.0,NaN,D,0.0,NaN,I,By Subsidiary
65171,0001493152-26-006706,8875287,Common Stock,11-FEB-2026,NaN,4.0,S,False,NaN,25000.0,NaN,D,712500.0,NaN,D,NaN
65172,0001493152-26-006705,8874167,Common Stock,12-FEB-2026,NaN,4.0,S,False,NaN,25000.0,NaN,D,217750.0,NaN,D,NaN


### Data Quality Note: Missing `TRANS_PRICEPERSHARE`

Of the 28,757 filtered transaction records, 58 (~0.2%) have a missing `TRANS_PRICEPERSHARE` value.

The missing values were investigated by `TRANS_CODE` and were not isolated to a single transaction type:

* **42 Purchases (`P`)**
* **16 Sales (`S`)**

A review of the available ownership fields showed that some missing-price purchase records were associated with indirect ownership structures, including family trusts and IRA rollovers, based on the `NATURE_OF_OWNERSHIP` field. The missing-price sales did not show a clear, consistent pattern in the available columns.

A definitive explanation for the missing prices would likely require reviewing the transaction footnotes. However, the `FOOTNOTES` file was intentionally outside the scope of this project because it was not required to answer the five defined business questions.

**Decision:** The 58 missing prices were left as `NaN` rather than imputed. There was no reliable basis in the available data to determine or estimate the actual transaction price without introducing unsupported assumptions.


## Cleaning function: process one quarter at a time

Wraps the full cleaning pipeline into one reusable function, `clean_quarter(quarter)`. <br>
The function reads the three files for a given quarter, standardizes boolean fields,<br> 
splits `RPTOWNER_RELATIONSHIP` into role flags, drops `_FN` columns, and filters transactions to purchases and sales (`P`/`S`).

The function is then called once for each quarter, avoiding the need to repeat the same cleaning logic 12 times.



In [49]:
def clean_quarter(quarter):
    # quarter is a placeholder — a string like "2026q1", handed in each time this function is called

    # Read the 3 files for this quarter (reuse your existing read logic, 
    # but swap the hardcoded "2026q1" for the quarter parameter instead)
    submission = pd.read_csv(f"raw_data/{quarter}/SUBMISSION.tsv", sep='\t', dtype={'ISSUERCIK': str})
    reportingowner = pd.read_csv(f'raw_data/{quarter}/REPORTINGOWNER.tsv', sep='\t', dtype={'RPTOWNERCIK': str})
    nonderiv_trans = pd.read_csv(f'raw_data/{quarter}/NONDERIV_TRANS.tsv', sep='\t')

    # ... apply your existing cleaning steps here (booleans, role flags, _FN drop, P/S filter) ...
    submission['NOT_SUBJECT_SEC16'] = submission['NOT_SUBJECT_SEC16'].replace({
        'false' : False,
        0 : False,
        '0' : False,
        'true' : True,
        1 : True,
        '1' : True
    })

    submission['AFF10B5ONE'] = submission['AFF10B5ONE'].replace({
        'false' : False,
        '0' : False,
        'true' : True,
        '1' : True
    })

    nonderiv_trans['EQUITY_SWAP_INVOLVED'] = nonderiv_trans['EQUITY_SWAP_INVOLVED'].replace({
        'false' : False,
        '0': False,
        '1' : True
    })

    nonderiv_trans = nonderiv_trans[nonderiv_trans['TRANS_CODE'].isin(['P', 'S'])]


    reportingowner['is_officer'] = reportingowner['RPTOWNER_RELATIONSHIP'].str.contains('Officer').astype('boolean')
    reportingowner.loc[reportingowner['RPTOWNER_RELATIONSHIP'].isna(), 'is_officer'] = pd.NA

    reportingowner['is_director'] = reportingowner['RPTOWNER_RELATIONSHIP'].str.contains('Director').astype('boolean')
    reportingowner.loc[reportingowner['RPTOWNER_RELATIONSHIP'].isna(), 'is_director'] = pd.NA

    reportingowner['is_ten_percent_owner'] = reportingowner['RPTOWNER_RELATIONSHIP'].str.contains('TenPercentOwner').astype('boolean')
    reportingowner.loc[reportingowner['RPTOWNER_RELATIONSHIP'].isna(), 'is_ten_percent_owner'] = pd.NA

    reportingowner['is_other'] = reportingowner['RPTOWNER_RELATIONSHIP'].str.contains('Other').astype('boolean')
    reportingowner.loc[reportingowner['RPTOWNER_RELATIONSHIP'].isna(), 'is_other'] = pd.NA

    fn_columns = nonderiv_trans.columns[nonderiv_trans.columns.str.endswith('_FN')]
    nonderiv_trans = nonderiv_trans.drop(columns=fn_columns)

    fn_columns = submission.columns[submission.columns.str.endswith('_FN')]
    submission = submission.drop(columns=fn_columns)

    fn_columns = reportingowner.columns[reportingowner.columns.str.endswith('_FN')]
    reportingowner = reportingowner.drop(columns=fn_columns)

    # Hand back all 3 cleaned tables so whoever calls this function can use them
    return submission, reportingowner, nonderiv_trans


In [50]:
# Test the function on one quarter before looping across all 12
test_submission, test_reportingowner, test_nonderiv_trans = clean_quarter('2026q1')

print(test_submission.shape)
print(test_reportingowner.shape)
print(test_nonderiv_trans.shape)

C:\Users\stephanie.mitchell\AppData\Local\Temp\ipykernel_15272\126082999.py:8: DtypeWarning: Columns (0: VALU_OWND_FOLWNG_TRANS_FN) have mixed types. Specify dtype option on import or set low_memory=False.
  nonderiv_trans = pd.read_csv(f'raw_data/{quarter}/NONDERIV_TRANS.tsv', sep='\t')


(69259, 14)
(72211, 17)
(28757, 16)
